src/middlewares/fetch.js

```diff
@@ -0,0 +1,97 @@
+function serialize(obj) {
+  const str = [];
+  Object.values(obj).filter((x) => {
+    if (Object.prototype.hasOwnProperty.call(obj, x)) {
+      str.push(`${encodeURIComponent(x)}=${encodeURIComponent(obj[x])}`);
+    }
+    return x;
+  });
+  return str.join('&');
+}
+
+const goTo = (action, store, next, payload, meta) => {
+  if (typeof action !== 'function') {
+    /* eslint-disable no-console */
+    console.error('fetchMiddleware require a function for onStart, onSuccess, onFailure');
+    /* eslint-enable no-console */
+    return;
+  }
+
+  const res = action(payload, meta, store.dispatch, store.getState);
+  if (res) {
+    next(res);
+  }
+};
+
+let config = {
+  base: '',
+  onRequest: null,
+  defaultParams: null,
+};
+
+const fetchMiddleware = store => next => (action) => {
+  if (!action.url) {
+    return next(action);
+  }
+  let request = {
+    url: action.url,
+    method: action.method || 'GET',
+    headers: action.headers,
+    body: action.body,
+    mode: action.mode || 'cors',
+  };
+
+  if (config.defaultParams || action.params) {
+    request.params = { ...config.defaultParams, ...(action.params || {}) };
+  }
+
+  if (config.onRequest) {
+    request = { ...request, ...config.onRequest(request, store.getState(), action) };
+  }
+
+  if (action.onStart) {
+    goTo(action.onStart, store, next);
+  }
+
+  const baseUrl = action.base ? action.base : config.base;
+  const params = request.params ? `?${serialize(request.params)}` : '';
+  fetch(`${baseUrl}${request.url}${params}`, {
```

Comment:
To write less code, you can also destructure request, like this:
const { headers, body, method, mode  } = request;
const body = JSON.stringify(body);

fetch(..., { body, headers, method, mode })